File: postgresql-async/src/test/scala/com/github/mauricio/async/db/postgresql/parsers/ParserESpec.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.postgresql.parsers

import com.github.mauricio.async.db.postgresql.messages.backend.{ServerMessage, ErrorMessage}
import org.jboss.netty.buffer.ChannelBuffers
import org.jboss.netty.util.CharsetUtil
import org.specs2.mutable.Specification

class ParserESpec extends Specification {

  "ErrorParser" should {

    "correctly parse an error message" in {

      val content = "this is my error message"
      val error = content.getBytes(CharsetUtil.UTF_8)
      val buffer = ChannelBuffers.dynamicBuffer()
      buffer.writeByte('M')
      buffer.writeBytes(error)
      buffer.writeByte(0)

      val message = new ErrorParser(CharsetUtil.UTF_8).parseMessage(buffer).asInstanceOf[ErrorMessage]

      message.message === content
      message.kind === ServerMessage.Error
    }

  }

}

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/parsers/ParserESpec.kt
package com.github.aysnc.sql.db.postgresql.parsers

import com.github.jasync.sql.db.postgresql.messages.backend.ErrorMessage
import com.github.jasync.sql.db.postgresql.messages.backend.ServerMessage
import com.github.jasync.sql.db.postgresql.parsers.ErrorParser
import io.netty.buffer.Unpooled
import io.netty.util.CharsetUtil
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class ParserESpec {

    @Test
    fun `ErrorParser should correctly parse an error message`() {
        val content = "this is my error message"
        val error = content.toByteArray(CharsetUtil.UTF_8)
        val buffer = Unpooled.buffer()
        buffer.writeByte('M'.toInt())
        buffer.writeBytes(error)
        buffer.writeByte(0)

        val message = ErrorParser(CharsetUtil.UTF_8).parseMessage(buffer) as ErrorMessage

        assertThat(message.message).isEqualTo(content)
        assertThat(message.kind).isEqualTo(ServerMessage.Error)
    }
}
